File: apps/services/shared/errors/exceptions.ts
/**
 * Custom Exception Classes
 * فئات الاستثناءات المخصصة
 *
 * @module shared/errors
 * @description Custom exception classes with bilingual support
 */

import { HttpException, HttpStatus } from '@nestjs/common';
import { ErrorCode, ERROR_REGISTRY } from './error-codes';

/**
 * Base Application Exception
 * الاستثناء الأساسي للتطبيق
 */
export class AppException extends HttpException {
  public readonly errorCode: ErrorCode;
  public readonly messageEn: string;
  public readonly messageAr: string;
  public readonly retryable: boolean;
  public readonly details?: any;
  public readonly timestamp: string;
  public readonly path?: string;

  constructor(
    errorCode: ErrorCode,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    const metadata = ERROR_REGISTRY[errorCode];
    const messageEn = customMessage?.en || metadata.message.en;
    const messageAr = customMessage?.ar || metadata.message.ar;

    super(
      {
        errorCode,
        message: messageEn,
        messageAr,
        details,
        timestamp: new Date().toISOString(),
      },
      metadata.httpStatus,
    );

    this.errorCode = errorCode;
    this.messageEn = messageEn;
    this.messageAr = messageAr;
    this.retryable = metadata.retryable;
    this.details = details;
    this.timestamp = new Date().toISOString();

    // Set prototype explicitly for instanceof checks
    Object.setPrototypeOf(this, AppException.prototype);
  }

  /**
   * Get error response object
   * الحصول على كائن استجابة الخطأ
   */
  toJSON() {
    return {
      success: false,
      error: {
        code: this.errorCode,
        message: this.messageEn,
        messageAr: this.messageAr,
        retryable: this.retryable,
        details: this.details,
        timestamp: this.timestamp,
        path: this.path,
      },
    };
  }
}

/**
 * Validation Exception
 * استثناء التحقق من صحة البيانات
 */
export class ValidationException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.VALIDATION_ERROR,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, ValidationException.prototype);
  }

  /**
   * Create validation exception from field errors
   * إنشاء استثناء التحقق من أخطاء الحقول
   */
  static fromFieldErrors(
    fieldErrors: Array<{ field: string; message: string; messageAr?: string }>,
  ): ValidationException {
    return new ValidationException(ErrorCode.VALIDATION_ERROR, undefined, {
      fields: fieldErrors,
    });
  }
}

/**
 * Authentication Exception
 * استثناء المصادقة
 */
export class AuthenticationException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.AUTHENTICATION_FAILED,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, AuthenticationException.prototype);
  }
}

/**
 * Authorization Exception
 * استثناء التفويض
 */
export class AuthorizationException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.FORBIDDEN,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, AuthorizationException.prototype);
  }
}

/**
 * Not Found Exception
 * استثناء عدم العثور على المورد
 */
export class NotFoundException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.RESOURCE_NOT_FOUND,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, NotFoundException.prototype);
  }

  /**
   * Create exception for specific resource types
   * إنشاء استثناء لأنواع موارد محددة
   */
  static user(userId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.USER_NOT_FOUND, undefined, { userId });
  }

  static farm(farmId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.FARM_NOT_FOUND, undefined, { farmId });
  }

  static field(fieldId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.FIELD_NOT_FOUND, undefined, { fieldId });
  }

  static crop(cropId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.CROP_NOT_FOUND, undefined, { cropId });
  }

  static sensor(sensorId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.SENSOR_NOT_FOUND, undefined, { sensorId });
  }

  static conversation(conversationId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.CONVERSATION_NOT_FOUND, undefined, {
      conversationId,
    });
  }

  static message(messageId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.MESSAGE_NOT_FOUND, undefined, {
      messageId,
    });
  }

  static wallet(walletId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.WALLET_NOT_FOUND, undefined, {
      walletId,
    });
  }

  static order(orderId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.ORDER_NOT_FOUND, undefined, { orderId });
  }

  static product(productId?: string): NotFoundException {
    return new NotFoundException(ErrorCode.PRODUCT_NOT_FOUND, undefined, {
      productId,
    });
  }
}

/**
 * Conflict Exception
 * استثناء التعارض
 */
export class ConflictException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.RESOURCE_ALREADY_EXISTS,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, ConflictException.prototype);
  }
}

/**
 * Business Logic Exception
 * استثناء منطق الأعمال
 */
export class BusinessLogicException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.BUSINESS_RULE_VIOLATION,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, BusinessLogicException.prototype);
  }

  /**
   * Create exception for specific business rules
   * إنشاء استثناء لقواعد عمل محددة
   */
  static insufficientBalance(
    available: number,
    required: number,
  ): BusinessLogicException {
    return new BusinessLogicException(ErrorCode.INSUFFICIENT_BALANCE, undefined, {
      available,
      required,
    });
  }

  static amountMustBePositive(amount: number): BusinessLogicException {
    return new BusinessLogicException(ErrorCode.AMOUNT_MUST_BE_POSITIVE, undefined, {
      amount,
    });
  }

  static invalidStateTransition(
    currentState: string,
    targetState: string,
  ): BusinessLogicException {
    return new BusinessLogicException(
      ErrorCode.INVALID_STATE_TRANSITION,
      undefined,
      {
        currentState,
        targetState,
      },
    );
  }

  static operationNotAllowed(
    operation: string,
    reason?: string,
  ): BusinessLogicException {
    return new BusinessLogicException(ErrorCode.OPERATION_NOT_ALLOWED, undefined, {
      operation,
      reason,
    });
  }
}

/**
 * External Service Exception
 * استثناء الخدمة الخارجية
 */
export class ExternalServiceException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.EXTERNAL_SERVICE_ERROR,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, ExternalServiceException.prototype);
  }

  /**
   * Create exception for specific external services
   * إنشاء استثناء لخدمات خارجية محددة
   */
  static weatherService(error?: any): ExternalServiceException {
    return new ExternalServiceException(
      ErrorCode.WEATHER_SERVICE_UNAVAILABLE,
      undefined,
      { originalError: error },
    );
  }

  static satelliteService(error?: any): ExternalServiceException {
    return new ExternalServiceException(
      ErrorCode.SATELLITE_SERVICE_UNAVAILABLE,
      undefined,
      { originalError: error },
    );
  }

  static paymentGateway(error?: any): ExternalServiceException {
    return new ExternalServiceException(
      ErrorCode.PAYMENT_GATEWAY_ERROR,
      undefined,
      { originalError: error },
    );
  }

  static smsService(error?: any): ExternalServiceException {
    return new ExternalServiceException(ErrorCode.SMS_SERVICE_ERROR, undefined, {
      originalError: error,
    });
  }

  static emailService(error?: any): ExternalServiceException {
    return new ExternalServiceException(ErrorCode.EMAIL_SERVICE_ERROR, undefined, {
      originalError: error,
    });
  }
}

/**
 * Database Exception
 * استثناء قاعدة البيانات
 */
export class DatabaseException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.DATABASE_ERROR,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, DatabaseException.prototype);
  }

  /**
   * Create exception from Prisma/TypeORM error
   * إنشاء استثناء من خطأ Prisma/TypeORM
   */
  static fromDatabaseError(error: any): DatabaseException {
    // Prisma error codes
    if (error.code === 'P2002') {
      return new DatabaseException(ErrorCode.UNIQUE_CONSTRAINT_VIOLATION, undefined, {
        fields: error.meta?.target,
      });
    }
    if (error.code === 'P2003') {
      return new DatabaseException(ErrorCode.FOREIGN_KEY_VIOLATION, undefined, {
        field: error.meta?.field_name,
      });
    }
    if (error.code === 'P2025') {
      return new DatabaseException(ErrorCode.RESOURCE_NOT_FOUND, undefined, {
        cause: error.meta?.cause,
      });
    }

    // Generic database error
    return new DatabaseException(ErrorCode.DATABASE_ERROR, undefined, {
      originalError: error.message,
    });
  }
}

/**
 * Internal Server Exception
 * استثناء الخادم الداخلي
 */
export class InternalServerException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.INTERNAL_SERVER_ERROR,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, InternalServerException.prototype);
  }
}

/**
 * Rate Limit Exception
 * استثناء تجاوز الحد المسموح
 */
export class RateLimitException extends AppException {
  constructor(
    errorCode: ErrorCode = ErrorCode.RATE_LIMIT_EXCEEDED,
    customMessage?: { en?: string; ar?: string },
    details?: any,
  ) {
    super(errorCode, customMessage, details);
    Object.setPrototypeOf(this, RateLimitException.prototype);
  }

  /**
   * Create exception with retry information
   * إنشاء استثناء مع معلومات إعادة المحاولة
   */
  static withRetryAfter(retryAfterSeconds: number): RateLimitException {
    return new RateLimitException(ErrorCode.RATE_LIMIT_EXCEEDED, undefined, {
      retryAfter: retryAfterSeconds,
      retryAfterDate: new Date(
        Date.now() + retryAfterSeconds * 1000,
      ).toISOString(),
    });
  }
}

File: apps/services/shared/errors_py/exceptions.py
"""
Custom Exception Classes
فئات الاستثناءات المخصصة

@module shared/errors_py
@description Custom exception classes with bilingual support
"""

from datetime import datetime
from typing import Any, Dict, Optional
from .error_codes import ErrorCode, ERROR_REGISTRY, BilingualMessage


class AppException(Exception):
    """
    Base Application Exception
    الاستثناء الأساسي للتطبيق
    """

    def __init__(
        self,
        error_code: ErrorCode,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        metadata = ERROR_REGISTRY[error_code]

        if custom_message:
            self.message_en = custom_message.en
            self.message_ar = custom_message.ar
        else:
            self.message_en = metadata.message.en
            self.message_ar = metadata.message.ar

        self.error_code = error_code
        self.http_status = metadata.http_status
        self.retryable = metadata.retryable
        self.details = details or {}
        self.timestamp = datetime.utcnow().isoformat()
        self.path: Optional[str] = None

        super().__init__(self.message_en)

    def to_dict(self) -> Dict[str, Any]:
        """
        Get error response object
        الحصول على كائن استجابة الخطأ
        """
        return {
            "success": False,
            "error": {
                "code": self.error_code.value,
                "message": self.message_en,
                "messageAr": self.message_ar,
                "retryable": self.retryable,
                "details": self.details if self.details else None,
                "timestamp": self.timestamp,
                "path": self.path,
            },
        }


class ValidationException(AppException):
    """
    Validation Exception
    استثناء التحقق من صحة البيانات
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.VALIDATION_ERROR,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)

    @classmethod
    def from_field_errors(
        cls,
        field_errors: list[Dict[str, str]],
    ) -> "ValidationException":
        """
        Create validation exception from field errors
        إنشاء استثناء التحقق من أخطاء الحقول
        """
        return cls(
            ErrorCode.VALIDATION_ERROR,
            details={"fields": field_errors},
        )


class AuthenticationException(AppException):
    """
    Authentication Exception
    استثناء المصادقة
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.AUTHENTICATION_FAILED,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)


class AuthorizationException(AppException):
    """
    Authorization Exception
    استثناء التفويض
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.FORBIDDEN,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)


class NotFoundException(AppException):
    """
    Not Found Exception
    استثناء عدم العثور على المورد
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.RESOURCE_NOT_FOUND,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)

    @classmethod
    def user(cls, user_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for user not found"""
        return cls(ErrorCode.USER_NOT_FOUND, details={"userId": user_id})

    @classmethod
    def farm(cls, farm_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for farm not found"""
        return cls(ErrorCode.FARM_NOT_FOUND, details={"farmId": farm_id})

    @classmethod
    def field(cls, field_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for field not found"""
        return cls(ErrorCode.FIELD_NOT_FOUND, details={"fieldId": field_id})

    @classmethod
    def crop(cls, crop_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for crop not found"""
        return cls(ErrorCode.CROP_NOT_FOUND, details={"cropId": crop_id})

    @classmethod
    def sensor(cls, sensor_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for sensor not found"""
        return cls(ErrorCode.SENSOR_NOT_FOUND, details={"sensorId": sensor_id})

    @classmethod
    def conversation(cls, conversation_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for conversation not found"""
        return cls(ErrorCode.CONVERSATION_NOT_FOUND, details={"conversationId": conversation_id})

    @classmethod
    def message(cls, message_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for message not found"""
        return cls(ErrorCode.MESSAGE_NOT_FOUND, details={"messageId": message_id})

    @classmethod
    def wallet(cls, wallet_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for wallet not found"""
        return cls(ErrorCode.WALLET_NOT_FOUND, details={"walletId": wallet_id})

    @classmethod
    def order(cls, order_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for order not found"""
        return cls(ErrorCode.ORDER_NOT_FOUND, details={"orderId": order_id})

    @classmethod
    def product(cls, product_id: Optional[str] = None) -> "NotFoundException":
        """Create exception for product not found"""
        return cls(ErrorCode.PRODUCT_NOT_FOUND, details={"productId": product_id})


class ConflictException(AppException):
    """
    Conflict Exception
    استثناء التعارض
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.RESOURCE_ALREADY_EXISTS,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)


class BusinessLogicException(AppException):
    """
    Business Logic Exception
    استثناء منطق الأعمال
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.BUSINESS_RULE_VIOLATION,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)

    @classmethod
    def insufficient_balance(
        cls,
        available: float,
        required: float,
    ) -> "BusinessLogicException":
        """Create exception for insufficient balance"""
        return cls(
            ErrorCode.INSUFFICIENT_BALANCE,
            details={"available": available, "required": required},
        )

    @classmethod
    def amount_must_be_positive(cls, amount: float) -> "BusinessLogicException":
        """Create exception for invalid amount"""
        return cls(
            ErrorCode.AMOUNT_MUST_BE_POSITIVE,
            details={"amount": amount},
        )

    @classmethod
    def invalid_state_transition(
        cls,
        current_state: str,
        target_state: str,
    ) -> "BusinessLogicException":
        """Create exception for invalid state transition"""
        return cls(
            ErrorCode.INVALID_STATE_TRANSITION,
            details={"currentState": current_state, "targetState": target_state},
        )

    @classmethod
    def operation_not_allowed(
        cls,
        operation: str,
        reason: Optional[str] = None,
    ) -> "BusinessLogicException":
        """Create exception for operation not allowed"""
        return cls(
            ErrorCode.OPERATION_NOT_ALLOWED,
            details={"operation": operation, "reason": reason},
        )


class ExternalServiceException(AppException):
    """
    External Service Exception
    استثناء الخدمة الخارجية
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.EXTERNAL_SERVICE_ERROR,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)

    @classmethod
    def weather_service(cls, error: Optional[Any] = None) -> "ExternalServiceException":
        """Create exception for weather service error"""
        return cls(
            ErrorCode.WEATHER_SERVICE_UNAVAILABLE,
            details={"originalError": str(error)} if error else None,
        )

    @classmethod
    def satellite_service(cls, error: Optional[Any] = None) -> "ExternalServiceException":
        """Create exception for satellite service error"""
        return cls(
            ErrorCode.SATELLITE_SERVICE_UNAVAILABLE,
            details={"originalError": str(error)} if error else None,
        )

    @classmethod
    def payment_gateway(cls, error: Optional[Any] = None) -> "ExternalServiceException":
        """Create exception for payment gateway error"""
        return cls(
            ErrorCode.PAYMENT_GATEWAY_ERROR,
            details={"originalError": str(error)} if error else None,
        )

    @classmethod
    def sms_service(cls, error: Optional[Any] = None) -> "ExternalServiceException":
        """Create exception for SMS service error"""
        return cls(
            ErrorCode.SMS_SERVICE_ERROR,
            details={"originalError": str(error)} if error else None,
        )

    @classmethod
    def email_service(cls, error: Optional[Any] = None) -> "ExternalServiceException":
        """Create exception for email service error"""
        return cls(
            ErrorCode.EMAIL_SERVICE_ERROR,
            details={"originalError": str(error)} if error else None,
        )


class DatabaseException(AppException):
    """
    Database Exception
    استثناء قاعدة البيانات
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.DATABASE_ERROR,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)

    @classmethod
    def from_database_error(cls, error: Any) -> "DatabaseException":
        """
        Create exception from database error
        إنشاء استثناء من خطأ قاعدة البيانات
        """
        error_str = str(error).lower()

        # Check for specific database errors
        if "unique constraint" in error_str or "duplicate key" in error_str:
            return cls(ErrorCode.UNIQUE_CONSTRAINT_VIOLATION)
        elif "foreign key" in error_str:
            return cls(ErrorCode.FOREIGN_KEY_VIOLATION)
        elif "timeout" in error_str:
            return cls(ErrorCode.QUERY_TIMEOUT)
        elif "connection" in error_str:
            return cls(ErrorCode.DATABASE_CONNECTION_FAILED)
        else:
            return cls(
                ErrorCode.DATABASE_ERROR,
                details={"originalError": str(error)},
            )


class InternalServerException(AppException):
    """
    Internal Server Exception
    استثناء الخادم الداخلي
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.INTERNAL_SERVER_ERROR,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)


class RateLimitException(AppException):
    """
    Rate Limit Exception
    استثناء تجاوز الحد المسموح
    """

    def __init__(
        self,
        error_code: ErrorCode = ErrorCode.RATE_LIMIT_EXCEEDED,
        custom_message: Optional[BilingualMessage] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(error_code, custom_message, details)

    @classmethod
    def with_retry_after(cls, retry_after_seconds: int) -> "RateLimitException":
        """
        Create exception with retry information
        إنشاء استثناء مع معلومات إعادة المحاولة
        """
        retry_after_date = datetime.utcnow()
        retry_after_date = retry_after_date.replace(
            second=retry_after_date.second + retry_after_seconds
        )

        return cls(
            ErrorCode.RATE_LIMIT_EXCEEDED,
            details={
                "retryAfter": retry_after_seconds,
                "retryAfterDate": retry_after_date.isoformat(),
            },
        )
